File: 219.存在重复元素-ii.py
#
# @lc app=leetcode.cn id=219 lang=python3
#
# [219] 存在重复元素 II
#

# @lc code=start
class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        # 以空间换时间
        all_dict = {}
        for i, v in enumerate(nums):
            # 从前往后，直接相减就是 abs
            if v in all_dict and i - all_dict[v] <= k:
                return True
            all_dict[v] = i
        return False
# @lc code=end

File: 219.存在重复元素-ii.go
/*
 * @lc app=leetcode.cn id=219 lang=golang
 *
 * [219] 存在重复元素 II
 */

// @lc code=start
func containsNearbyDuplicate(nums []int, k int) bool {
	allMap := map[int]int{}
	for idx, val := range nums {
		if preIndex, ok := allMap[val]; ok {
			if idx-preIndex <= k {
				return true
			}
		}
		// 即使重复也希望保留最近的数字，因为要求小于，所以离得近值自然越小
		allMap[val] = idx
	}
	// 不满足条件
	return false
}

// @lc code=end

